subsys/bluetooth/mesh/light_ctrl_srv.c

```diff
@@ -0,0 +1,1449 @@
+}
+
+static bool state_is_on(const struct bt_mesh_light_ctrl_srv *srv,
+			enum bt_mesh_light_ctrl_srv_state state)
+{
+	/* Only the stable Standby state is considered Off: */
+	if (srv->fade.duration > 0) {
+		return (state != LIGHT_LC_STATE_STANDBY ||
+			atomic_test_bit(&srv->flags, FLAG_TRANSITION));
+	}
+
+	return srv->state != LIGHT_LC_STATE_STANDBY;
+}
+
+static void onoff_encode(struct bt_mesh_light_ctrl_srv *srv,
+			 struct net_buf_simple *buf,
+			 enum bt_mesh_light_ctrl_srv_state prev_state)
+{
+	bt_mesh_model_msg_init(buf, BT_MESH_LIGHT_CTRL_OP_LIGHT_ONOFF_STATUS);
+	net_buf_simple_add_u8(buf, state_is_on(srv, prev_state));
+
+	u32_t remaining_fade = remaining_fade_time(srv);
+
+	if (remaining_fade) {
+		net_buf_simple_add_u8(buf,
+				      srv->state != LIGHT_LC_STATE_STANDBY);
+		net_buf_simple_add_u8(buf,
+				      model_transition_encode(remaining_fade));
+		return;
+	}
+
+	if (atomic_test_bit(&srv->flags, FLAG_ON_PENDING) ||
+	    atomic_test_bit(&srv->flags, FLAG_OFF_PENDING)) {
+		remaining_fade =
+			k_delayed_work_remaining_get(&srv->action_delay) +
+			srv->fade.duration;
+		net_buf_simple_add_u8(buf, atomic_test_bit(&srv->flags,
+							   FLAG_ON_PENDING));
+		net_buf_simple_add_u8(buf,
+				      model_transition_encode(remaining_fade));
+		return;
+	}
+}
+
+static int onoff_status_send(struct bt_mesh_light_ctrl_srv *srv,
+			     struct bt_mesh_msg_ctx *ctx,
+			     enum bt_mesh_light_ctrl_srv_state prev_state)
+{
+	BT_MESH_MODEL_BUF_DEFINE(buf, BT_MESH_LIGHT_CTRL_OP_LIGHT_ONOFF_STATUS,
+				 3);
+	onoff_encode(srv, &buf, prev_state);
+
+	return model_send(srv->model, ctx, &buf);
+}
+
+static void onoff_pub(struct bt_mesh_light_ctrl_srv *srv,
+		      enum bt_mesh_light_ctrl_srv_state prev_state,
+		      bool pub_gen_onoff)
+{
+	onoff_status_send(srv, NULL, prev_state);
+
+	if (!pub_gen_onoff) {
+		return;
+	}
+
+	u32_t remaining_fade;
+
+	if (atomic_test_bit(&srv->flags, FLAG_ON_PENDING) ||
+	    atomic_test_bit(&srv->flags, FLAG_OFF_PENDING)) {
+		remaining_fade =
+			k_delayed_work_remaining_get(&srv->action_delay) +
+			srv->fade.duration;
+	} else {
+		remaining_fade = remaining_fade_time(srv);
+	}
+
+	const struct bt_mesh_onoff_status status = {
+		state_is_on(srv, prev_state),
+		state_is_on(srv, srv->state),
+		remaining_fade,
+	};
+
+	bt_mesh_onoff_srv_pub(&srv->onoff, NULL, &status);
+}
+
+static u16_t light_get(struct bt_mesh_light_ctrl_srv *srv)
+{
+	if (!is_enabled(srv)) {
+		return 0;
+	}
+
+	if (!atomic_test_bit(&srv->flags, FLAG_TRANSITION) ||
+	    !srv->fade.duration) {
+		return srv->cfg.light[srv->state];
+	}
+
+	u32_t curr = curr_fade_time(srv);
+	u16_t start = srv->fade.initial_light;
+	u16_t end = srv->cfg.light[srv->state];
+
+	/* Linear interpolation: */
+	return start + ((end - start) * curr) / srv->fade.duration;
+}
+
+static void lux_get(struct bt_mesh_light_ctrl_srv *srv,
+		    struct sensor_value *lux)
+{
+#if CONFIG_BT_MESH_LIGHT_CTRL_SRV_REG
```

Comment:
Could we use if (IS_ENABLED()) here instead? Easier to catch compilation issues in CI this way.
Move `#if CONFIG_BT_MESH_LIGHT_CTRL_SRV_REG` only around the places strictly needed.

Reply:
Improved a bit upon this, but as the regulator is conditionally left out of the struct, it's hard to use IS_ENABLED without adding some other mess. Discussed offline.